src/cart/crt_group.c

```diff
@@ -3370,6 +3370,55 @@ crt_register_eviction_cb(crt_eviction_cb cb, void *arg)
 	return rc;
 }
 
+int
+crt_register_event_cb(crt_event_cb event_handler, void *arg)
+{
+	struct crt_event_cb_priv	*cb_priv;
+	int				 rc = 0;
+
+	D_ALLOC_PTR(cb_priv);
+	if (cb_priv == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	cb_priv->cecp_func = event_handler;
+	cb_priv->cecp_args = arg;
+
+	D_RWLOCK_WRLOCK(&crt_plugin_gdata.cpg_event_rwlock);
+	d_list_for_each_entry(cb_priv, &crt_plugin_gdata.cpg_event_cbs,
+			      cecp_link) {
+		if (cb_priv->cecp_func == event_handler &&
+		    cb_priv->cecp_args == arg) {
+			D_FREE(cb_priv);
+			rc = -DER_EXIST;
+			break;
+		}
+	}
+	if (rc == 0)
+		d_list_add_tail(&cb_priv->cecp_link,
+				&crt_plugin_gdata.cpg_event_cbs);
+	D_RWLOCK_UNLOCK(&crt_plugin_gdata.cpg_event_rwlock);
+
+out:
+	return rc;
+}
+
+void
+crt_unregister_event_cb(crt_event_cb event_handler, void *arg)
+{
+	struct crt_event_cb_priv *cb_priv, *next;
+
+	D_RWLOCK_WRLOCK(&crt_plugin_gdata.cpg_event_rwlock);
+	d_list_for_each_entry_safe(cb_priv, next,
+				   &crt_plugin_gdata.cpg_event_cbs,
+				   cecp_link) {
+		if (cb_priv->cecp_func == event_handler &&
+		    cb_priv->cecp_args == arg) {
+			d_list_del(&cb_priv->cecp_link);
+			D_FREE(cb_priv);
```

Comment:
Presumably you could break here as this check should never pass twice?  It would mean you didn't need to use the _safe() version of the macro as well.